### ISSUE TITLE
Meaningless generated images during pre-training despite normal loss descent

### ISSUE THREAD

**Issue by mysticalwing** [Amshaker/Mobile-O#6]
Hi authors,

Thanks for sharing this amazing project! I am currently trying to reproduce the Pre-training (Cross-Modal Alignment) stage based on your open-sourced code.

I am using scripts/Mobile-O-0.5B/pre_train.sh with Sana as the diffusion model. However, I encountered a strange issue: after training for about 5k steps, the generated images are completely meaningless (mostly random color blocks or blurred textures).

What confuses me is that the training process seems very healthy from the metrics perspective. The diffusion loss and gradient norms are decreasing smoothly without any noticeable spikes or fluctuations.

I have already verified a few basic setups:
1. Sana uses v-prediction, and the Flow Matching loss / noisy latents calculation seems aligned.
2. Sana's VAE only has a scaling_factor (no shift_factor), which matches the logic in decode_latents.
3. The prompt embeddings passed to the diffusion_connector are pure text embeddings (the <image> placeholder is correctly removed during data preprocessing).

Since the loss is decreasing but the visuals are collapsed, I would like to ask for your advice on potential debugging directions:

Insufficient training steps? Is 5k steps simply too early for the DiT to align with the LLM's text embeddings? Around which step should I expect to see rough outlines or meaningful structures?

Missing CFG Dropout during training? I noticed that in mobileo_sft.py's forward function, the encoder_hidden_states are directly passed to the DiT without applying null-text dropout (e.g., the mask_drop function in llava_arch.py). Could this cause the model to collapse during inference when CFG (guidance_scale > 1.0) feeds it zeroed-out unconditional embeddings?

Are there any other specific configurations or potential pitfalls I should check? Any guidance or hints would be greatly appreciated. Thank you!

**Comment by Amshaker** [comment]
Hi @mysticalwing ,

Thank you for your interest in our work.

On the 5k steps question: yes, way too early. The original pre-training was 10x more! We have not deeply tested the performance after pre-training. We monitored the loss and did a deep analysis after SFT, actually. But I think that's totally normal. In our experience, you should start seeing meaningful structure somewhere around 25k–30k steps, and the 0.5B model needs the full pre-training run to produce coherent output. The fact that your loss is going down smoothly without spikes is actually the best signal you can have right now.

On the CFG dropout: we have not used it in training, and the inference was working fine with or even without CFG. But you could try using a dropout probability of 0.1 in your setup, so roughly 10% of samples would have their encoder_hidden_states replaced with zeros during the forward pass. The only drawback is that the model is already small, and this might affect the quality.

Also, make sure to print which components are learnable during pre-training and verify that it's as expected.

I hope it helps!

Thank you!

**Comment by mysticalwing** [comment]
Hi @Amshaker 

Thanks again for your earlier reply. I’ve been debugging this issue recently, and I think I’ve found the root cause on my side.

In the pre-train script, `is_train=True` is not passed. Since `is_train` defaults to `False`,  in https://github.com/Amshaker/Mobile-O/blob/1e1c4a2e146a04b100842c2eed7dad61d5a4725b/mobileo/model/llava_arch.py#L55-L65 the scheduler is initialized as `DPMSolverMultistepScheduler` instead of `FlowMatchEulerDiscreteScheduler`, which seems to be the correct one for training.

After explicitly setting `is_train=True`, the training behavior becomes normal. In my case, the model already starts to produce reasonably good images after around **2k** steps. Here is a example with prompt "a photo of a cute cat"

<img width="960" height="1004" alt="Image" src="https://github.com/user-attachments/assets/6255ee72-f08b-43cb-ae33-a8df86f026c8" />

So the meaningless generations I observed before were likely caused by using the wrong scheduler during training.

It might also be worth updating the `pre_train` script accordingly, since `is_train` is explicitly set in both the `sft` and `post_train` scripts.

Thanks again for your help, and I hope this is useful.

**Comment by Amshaker** [comment]
Hi @mysticalwing ,

Thank you for the thorough and insightful reply!

You're absolutely right, `FlowMatchEulerDiscreteScheduler` should be used during training rather than `DPMSolverMultistepScheduler`.  While `DPMSolverMultistepScheduler` works better in the inference stage.

I've revised the training scripts accordingly: `is_train` was already explicitly set in both the `sft` and `post_train` stages. It had been inadvertently omitted only in the `pre_train` script, which has been corrected now. 

We appreciate your careful reading of the work and your interest in the project!
